src/Sia.Gateway/Hubs/EventsHub.cs

```diff
@@ -14,15 +18,90 @@ namespace Sia.Gateway.Hubs
     public class EventsHub : Hub
     {
         public const string HubPath = "events/live";
-        public EventsHub() : base()
+        private readonly ConcurrentDictionary<string, IFilterByMatch<Event>> _filterLookup;
+        private readonly ILogger<EventsHub> _logger;
+
+        public EventsHub(
+            ConcurrentDictionary<string, IFilterByMatch<Event>> filterLookup,
+            ILoggerFactory loggerFactory
+        ) : base()
         {
+            _filterLookup = filterLookup;
+            _logger = loggerFactory.CreateLogger<EventsHub>();
+        }
 
+        public override async Task OnConnectedAsync()
+        {
+            await base.OnConnectedAsync();
+            _logger.LogInformation(
+                "User {0} connected to EventsHub on connectionId {1}",
+                new object[] { Context.User.Identity.Name, Context.ConnectionId }
+            );
+            await ClearFilter();
+        }
+
+        public override async Task OnDisconnectedAsync(Exception exception)
+        {
+            await ClearFilter();
+            
+            if(exception == null)
+            {
+                _logger.LogInformation(
+                    "User {0} disconnected from EventsHub on connectionId {1} without exception",
+                    new object[] { Context.User.Identity.Name, Context.ConnectionId }
+                );
+            }
+            else
+            {
+                _logger.LogError(
+                    exception,
+                    "User {0} disconnected from EventsHub on connectionId {1} with exception; see exception for details.",
+                    new object[] { Context.User.Identity.Name, Context.ConnectionId }
+                );
+            }
+
+            await base.OnDisconnectedAsync(exception);
         }
 
         public Task Send(Event ev)
+            => Clients.AllExcept(
+                    _filterLookup
+                        .Where((kvp) => !kvp.Value.IsMatchFor(ev))
+                        .Select((kvp) => kvp.Key)
+                        .Append(Context.ConnectionId) // Prevent loops (if those are even possible)
+                        .ToList()
+                ).InvokeAsync("Send", Json(ev));
+
+        public Task UpdateFilter(EventFilters filters)
+        {
+            _logger.LogInformation(
+                "UpdateFilter called from connection with Id {0}.",
+                new object[] { Context.ConnectionId });
+            _filterLookup.Upsert(Context.ConnectionId, filters);
+            //No need to propagate to other Gateways as connectionIds are not shared.
+            return Task.CompletedTask;
+        }
+
+        public Task<bool> ClearFilter()
         {
-            return Clients.All.InvokeAsync("Send", Json(ev));
+            var filterRemoved = _filterLookup.TryRemove(Context.ConnectionId, out var unused);
+            if(filterRemoved)
```

Comment:
Space between if and condition: 

`if(filterRemoved)` => `if (filterRemoved)`

Reply:
Fixed this.

Any idea how we can enforce style across the team (preferably using settings checked into source control)? We have a linter that checks for stuff like this in the JavaScript code base, but it would be nice to have something that works similarly for the C# code base.

Reply:
StyleCop Analyzers can do the trick. It's very opinionated about code formatting https://github.com/DotNetAnalyzers/StyleCopAnalyzers

For example, it finds all kinds of formatting problems with this: 
![image](https://user-images.githubusercontent.com/2158838/37484207-20bb7142-2845-11e8-97a7-a669fd291227.png)